Cython/Compiler/ExprNodes.py

```diff
@@ -6653,87 +6688,19 @@ def generate_evaluation_code(self, code):
-        reuse_function_temp = self.function.is_temp
-        if reuse_function_temp:
-            function = self.function.result()
-        else:
-            function = code.funcstate.allocate_temp(py_object_type, manage_ref=True)
-            self.function.make_owned_reference(code)
-            code.put("%s = %s; " % (function, self.function.py_result()))
-            self.function.generate_disposal_code(code)
-            self.function.free_temps(code)
-
-        self_arg = code.funcstate.allocate_temp(py_object_type, manage_ref=True)
-        code.putln("%s = NULL;" % self_arg)
-        arg_offset_cname = code.funcstate.allocate_temp(PyrexTypes.c_uint_type, manage_ref=False)
-        code.putln("%s = 0;" % arg_offset_cname)
-
-        def attribute_is_likely_method(attr):
-            obj = attr.obj
-            if obj.is_name and obj.entry.is_pyglobal:
-                return False  # more likely to be a function
-            return True
-
-        if self.function.is_attribute:
-            likely_method = 'likely' if attribute_is_likely_method(self.function) else 'unlikely'
-        elif self.function.is_name and self.function.cf_state:
-            # not an attribute itself, but might have been assigned from one (e.g. bound method)
-            for assignment in self.function.cf_state:
-                value = assignment.rhs
-                if value and value.is_attribute and value.obj.type and value.obj.type.is_pyobject:
-                    if attribute_is_likely_method(value):
-                        likely_method = 'likely'
-                        break
-            else:
-                likely_method = 'unlikely'
-        else:
-            likely_method = 'unlikely'
-
-        if self.unpack:
-            # unpack is ultimately governed by optimize.unpack_method_calls
-            # and is a separate decision to whether we want vectorcall-type behaviour
-            code.putln("#if CYTHON_UNPACK_METHODS")
-            code.putln("if (%s(PyMethod_Check(%s))) {" % (likely_method, function))
-            code.putln("%s = PyMethod_GET_SELF(%s);" % (self_arg, function))
-            # the result of PyMethod_GET_SELF is always true in Py3.
-            code.putln(f"assert({self_arg});")
-            code.putln("PyObject* function = PyMethod_GET_FUNCTION(%s);" % function)
-            code.put_incref(self_arg, py_object_type)
-            code.put_incref("function", py_object_type)
-            # free method object as early to possible to enable reuse from CPython's freelist
-            code.put_decref_set(function, py_object_type, "function")
-            code.putln("%s = 1;" % arg_offset_cname)
-            code.putln("}")
-            code.putln("#endif")  # CYTHON_UNPACK_METHODS
-            # TODO may need to deal with unused variables in the #else case
-
+    def generate_callargs(self, code, kwargs_key_value_pairs, args, self_arg):
```

Comment:
Having two separate variants (with/without `kwargs_key_value_pairs`) of this method should really simplify the implementations.